spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/RabbitTemplate.java

```diff
@@ -1159,16 +1192,33 @@ public Message doInRabbit(Channel channel) throws Exception {
 				if (logger.isDebugEnabled()) {
 					logger.debug("Sending message with tag " + messageTag);
 				}
-				doSend(channel, exchange, routingKey, message, correlationData);
-				LinkedBlockingQueue<Message> replyHandoff = pendingReply.getQueue();
-				Message reply = (replyTimeout < 0) ? replyHandoff.take() : replyHandoff.poll(replyTimeout,
-						TimeUnit.MILLISECONDS);
-				RabbitTemplate.this.replyHolder.remove(messageTag);
+				Message reply = null;
+				try {
+					reply = exchangeMessages(exchange, routingKey, message, correlationData, channel, pendingReply,
+							messageTag);
+				}
+				finally {
+					RabbitTemplate.this.replyHolder.remove(messageTag);
+				}
 				return reply;
 			}
 		}, obtainTargetConnectionFactoryIfNecessary(this.sendConnectionFactorySelectorExpression, message));
 	}
 
+	private Message exchangeMessages(final String exchange, final String routingKey, final Message message,
+			final CorrelationData correlationData, Channel channel, final PendingReply pendingReply, String messageTag)
+			throws Exception, InterruptedException {
+		Message reply;
```

Comment:
Minor, of  course, but anyway:
`InterruptedException` isn't thrown right now.
The `reply` is redundant.

Sorry, the solution isn't clear to me, so I try to simplify code to figure out.